src/core/dbus-manager.c

```diff
@@ -3187,6 +3217,11 @@ const sd_bus_vtable bus_manager_vtable[] = {
                                 SD_BUS_RESULT("h", fd),
                                 method_dump_units_matching_patterns_by_fd,
                                 SD_BUS_VTABLE_UNPRIVILEGED),
+        SD_BUS_METHOD_WITH_ARGS("DumpMemoryStateByFileDescriptor",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("h", fd),
+                                method_dump_memory_state_by_fd,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
```

Comment:
hmm, this is a glibc-specific interface we will end up making an official dbus API. not sure how I like that.

i.e. if we ever port any of this to rust and leave glibc (or in fact any libc) behind, what do you expect will happen with this?

At the absolute minimum this should have SD_BUS_VTABLE_HIDDEN. But quite frankly I am not convinced this should be part of the dbus interface at all.

I mentioned this before, but I think this would be better done via some other mechanism, i.e. varlink, or just logging to journal or so, i.e. something that makes this a lot less official and API relevant.

Reply:
also, i am not convinced this should even be on pid1's interface at all. it's not specific to pid 1 after all, it's something that might make sense on *any* process using glibc. It sounds wrong to add this specifically to pid1's dbus interface even though it is that generic.

hence, at the absolute minimum this should be on a separate dbus interface. (i.e. like the LogControl stuff). But quite frankly, this too smells as if we should focus on a different mechanism, i.e. maybe trigger this via a signal (let's say SIGRTMIN+17 with a sigqueue value of 0x44) and then just have it dump to syslog, and then implement that everywhere?

Reply:
> hmm, this is a glibc-specific interface we will end up making an official dbus API. not sure how I like that.

IIRC our Dump() methods are not guaranteed to be stable in the output, and can change as the implementation changes. This will be the same - if the implementation changes in any way, then the output will be different. I'll clarify that in the documentation. I don't think it's a big deal, as like the Dump() it's a debugging tool.

> At the absolute minimum this should have SD_BUS_VTABLE_HIDDEN.

Seems strange given the other Dump methods are not hidden, but ok

> I mentioned this before, but I think this would be better done via some other mechanism, i.e. varlink, or just logging to journal or so, i.e. something that makes this a lot less official and API relevant.

Those mechanism can also be added in addition, sure. But I do need a D-Bus interface for this. My clients do not use varlink, and it's hard enough to make them use D-Bus, so varlink is no good to me. And I do not want to clobber my journals, so that's also no good to me. Finally I want to know who's querying so I can go after them if they overdo it, and also I definitely do NOT want to give these clients permission to send signals to PID1. So I have no problem if those other interfaces get added separately, but I also need this one.

Reply:
There's more than dbus in the world. Just make them call `systemd-analyze` in some way?

Reply:
That doesn't let me know who's querying, and also shelling out is bad practice and led to a lot of issues and security holes, so we are making everyone use programmatic APIs instead from C++.

Reply:
> IIRC our Dump() methods are not guaranteed to be stable in the output, and can change as the implementation changes. This will be the same - if the implementation changes in any way, then the output will be different. I'll clarify that in the documentation. I don't think it's a big deal, as like the Dump() it's a debugging tool.

changing output is one thing, but the whole API will be useless in the rustified future...

Reply:
(and already for people who use musl?)

Reply:
> Seems strange given the other Dump methods are not hidden, but ok

They probably should be actually. But this one is even weaker than those, since it's not even our internal data we spew out here, but glibcs.

Reply:
There must be memory introspection functionality in rust, surely? And if not, at some point it will probably be added, given it's useful to be able to get this information. I don't know about musl, but we don't support musl for anything else, so don't see why it should be an issue for this

The name can be changed to be more generic or implementation agnostic, I don't really mind how it's called and I'm open to suggestions

Reply:
but musl/rust would return this in a very different format most likely. using a generic function that returns completely different stuff depending on the implementation is a bad api.